core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala

```diff
@@ -0,0 +1,371 @@
+    actorSystem.awaitTermination()
+  }
+
+  override def toString = s"${getClass.getSimpleName}($actorSystem)"
+}
+
+private[spark] object AkkaRpcEnv {
+
+  def apply(config: RpcEnvConfig): RpcEnv = {
+    val (actorSystem, boundPort) = AkkaUtils.createActorSystem(
+        config.name, config.host, config.port, config.conf, config.securityManager)
+    new AkkaRpcEnv(actorSystem, config.conf, boundPort)
+  }
+
+}
+
+private[akka] class AkkaRpcEndpointRef(
+    @transient defaultAddress: RpcAddress,
+    val actorRef: ActorRef,
+    @transient conf: SparkConf) extends RpcEndpointRef with Serializable with Logging {
+  // `defaultAddress` and `conf` won't be used after initialization. So it's safe to be transient.
+
+  private[this] val maxRetries = conf.getInt("spark.akka.num.retries", 3)
+  private[this] val retryWaitMs = conf.getLong("spark.akka.retry.wait", 3000)
+  private[this] val defaultTimeout = conf.getLong("spark.akka.lookupTimeout", 30) seconds
+
+  override val address: RpcAddress = {
+    val akkaAddress = actorRef.path.address
+    RpcAddress(akkaAddress.host.getOrElse(defaultAddress.host),
+      akkaAddress.port.getOrElse(defaultAddress.port))
+  }
+
+  override val name: String = actorRef.path.name
+
+  override def askWithReply[T: ClassTag](message: Any): T = askWithReply(message, defaultTimeout)
+
+  override def askWithReply[T: ClassTag](message: Any, timeout: FiniteDuration): T = {
+    // TODO: Consider removing multiple attempts
+    var attempts = 0
+    var lastException: Exception = null
+    while (attempts < maxRetries) {
+      attempts += 1
+      try {
+        val future = sendWithReply[T](message, timeout)
+        val result = Await.result(future, timeout)
+        if (result == null) {
+          throw new SparkException("Actor returned null")
+        }
+        return result
+      } catch {
+        case ie: InterruptedException => throw ie
+        case e: Exception =>
+          lastException = e
+          logWarning(s"Error sending message [message = $message] in $attempts attempts", e)
+      }
+      Thread.sleep(retryWaitMs)
+    }
+
+    throw new SparkException(
+      s"Error sending message [message = $message]", lastException)
+  }
+
+  override def send(message: Any): Unit = {
+    actorRef ! AkkaMessage(message, false)
+  }
+
+  override def sendWithReply(message: Any, sender: RpcEndpointRef): Unit = {
+    implicit val actorSender: ActorRef =
+      if (sender == null) {
+        Actor.noSender
+      } else {
+        require(sender.isInstanceOf[AkkaRpcEndpointRef])
+        sender.asInstanceOf[AkkaRpcEndpointRef].actorRef
+      }
+    actorRef ! AkkaMessage(message, true)
+  }
+
+
+  override def sendWithReply[T: ClassTag](message: Any): Future[T] = {
+    sendWithReply(message, defaultTimeout)
+  }
+
+  override def sendWithReply[T: ClassTag](message: Any, timeout: FiniteDuration): Future[T] = {
+    import scala.concurrent.ExecutionContext.Implicits.global
```

Comment:
The problem of using  Akka's actorSystem's ExecutionContext is that I cannot get `actorSystem/context` from `actorRef`.